Step2 - 연관관계매핑 리뷰요청드립니다!

aksgudwns opened the pull request:
step2 연관관계매핑 리뷰요청드립니다!
감사합니다!

aksgudwns committed:
doc - step2 요구사항 추가

aksgudwns committed:
feat - Question 연관관계 추가

aksgudwns committed:
refactor - constraint naming

aksgudwns committed:
feat - Answer 연관관계 추가

- Answer 연관관계 추가
- QuestionTest 네임 리팩토링 setUp -> init
- BaseTimeEntity 실제 구현할 일이 없으므로 abstract클래스로 변경
- @Repository 어노테이션 추가

aksgudwns committed:
feat - DeleteHistory 연관관계 추가

- DeleteHistory 연관관계추가
- 테스트클래스 접근자제어 추가 및 상수선언 추가

joojimin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -16,7 +16,9 @@ public class Answer extends BaseTimeEntity {
     private boolean deleted = false;
     private Long questionId;
-    private Long writerId;
+    @ManyToOne
```
`@ManyToOne`의 속성중에는 fetchType이라는게 존재하고 LAZY, EAGAR의 값을 가질 수 있는데요!! 해당기능을 어떤기능을 할까요?? 또 저희코드에서 적용되면 좋을 속성은 무엇일까요!

aksgudwns replied:
LAZY는 지연로딩 방식이고, EAGER는 즉시로딩 방식입니다
LAZY는 프록시 객체를 먼저 조회만 하고, 객체의 필드 값을 조회할 때 쿼리가 나가는 방식입니다. 
따라서 객체에 있는 필드 값을 조회하는 비즈니스 로직이 많으면 EAGER방식이 유리하고, 반대면 LAZY방식이 유리하다고 알고 있습니다!

저희코드를 보니 Answer에서 User와 Question, Question에서 User의 필드를 접근하는 비즈니스 로직이 없기 때문에 LAZY방식이 유리하다고 생각됩니다.

추가 반영하겠습니다!

joojimin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -13,7 +15,12 @@ public class Question extends BaseTimeEntity {
+    @ManyToOne
+    @JoinColumn(name = "writer_id", foreignKey = @ForeignKey(name = "fk_question_writer"))
+    private User writer;
+    @OneToMany
```
관계형 데이터베이스 속성을 고민해봐야할 거 같아요~!
의도하신바가 외래키 주인설정을 일대다 관계에서 일쪽에 설정하신게 맞으실까요~?
혹시 맞다면 이유를 듣고싶습니다~!

aksgudwns replied:
제가 억지로 단방향 설정으로 풀려다 보니, 그렇게 되었습니다..
양방향 설정으로 다시 작성하겠습니다!

joojimin reviewed (changes requested):
안녕하세요~ 형준님!
미션진행 고생하셨습니다.
몇가지 코멘트 남겼는데 확인해보시고 리뷰 재요청 부탁드려요!!

추가로 QnaService에서도 데이터중심의 코드에서 객체 지향인 코드로 변경하면 좋을 것 같습니다!
미션의 힌트를 잘 참고해주세요~!

aksgudwns committed:
refactor - 데이터 중심 코드에서 객체지향 코드로 변경

- OneToMany에 외래키 설정한 부분 양방향 설정으로 수정
- fetchType Lazy타입 추가
- 데이터 중심에서 객체지향 코드로 리팩토링  DeleteHistory 생성자, AnswerRepository 질문탐색 메소드, 이에 따른 QnaService 비즈니스로직 수정

aksgudwns commented:
코멘트 감사합니다! 
말씀주신 내용 반영해보았습니다..

실무에서 안써서 그런지 JPA 아직 더 많이 공부 해야겠다는 생각이드네요..!
그래도 과제 하면서 많이 배우는 느낌들어서 좋습니다 :)

제가 잘못 이해 했거나 보완할 부분있다면 피드백 부탁드리겠습니다!

joojimin commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -39,7 +39,7 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         }
 
-        List<Answer> answers = answerRepository.findByQuestionIdAndDeletedFalse(questionId);
+        List<Answer> answers = question.getAnswers();
```
객체 지향적인 코드로 변경 잘해주셨습니다!
추가로 코멘트를 드려야할거같아요 ㅠㅠ 전부터 연습하고있는 일급컬렉션을 사용하면 좋을 것 같습니다!
아래 로직들도 깔끔하게 정리될거같아요

joojimin reviewed (changes requested):
피드백 반영 고생하셨습니다 형준님!
다만 추가적으로 코멘트 한가지 드려봤어요!
해당 내용 확인해보시고 리뷰 재요청 부탁드립니다!

aksgudwns committed:
refactor - 일급콜렉션 사용 추가 및 서비스 비즈니스로직 엔터티로 이동

 - Answers 일급콜렉션 추가
 - getAnswers 일급콜렉션 리턴하도록 변경
 - 서비스에 있는 비즈니스로직 엔터티로 이동

aksgudwns commented:
빠른 피드백 주셔서 감사합니다!

일급컬렉션 사용하면서 서비스에 있는 비즈니스 로직도 엔터티로 옮겨서 수정해보았습니다!
리뷰부탁드리겠습니다..!

감사합니다.

joojimin reviewed (approved):
피드백 반영 고생하셨습니다! 형준님~!
코드가 많이 깔끔해졌네요!! 다음미션도 화이팅입니다~!!
